### ISSUE TITLE
Chrome to throttle background tabs

### ISSUE THREAD

**Issue by Carreau** [jupyterlab/jupyterlab#1545]
cf http://blog.strml.net/2017/01/chrome-56-now-aggressively-throttles.html

TL:DR; Background tabs will only get 0.01s of CPU time/sec

I'm going to guess this can likely affect JupyterLab (and notebook), more if there are widgets on the page, and/or anything that receives messages from the kernel.

That _may_ explain, why my chrome (I'm on 56 beta), regularly seem to "Freeze" when I come on a page and suddenly speedup. This I think include JupyterLab though I wouldn't swear, and Chrome Beta has in general had a few other quirks, so it _may_ be something else.

[Edit] It is active in 56-beta but is planned to ship on 57, there are comments on HN pointing that the budget allocating to tabs can depends on websockets usage, playing audio...[/edit]

**Comment by blink1073** [comment]
Thanks for the heads up!  I'll mark this as a 1.0 issue so we at least check out the ramifications.

**Comment by jasongrout** [comment]
(1) my understanding is that an open websocket connection disables this optimization, and (2) we use requestAnimateFrame quite a bit right now, which I understand may do this already.

**Comment by sccolbert** [comment]
Also, not sure what our recourse would be here. We can't force the browser to *not* throttle.

**Comment by Carreau** [comment]
> Also, not sure what our recourse would be here. We can't force the browser to not throttle.

I never said we should find a way for the browser to not throttle, but when not on focus you can make sure to do only the minimum amount of work;  queue what could be expensive until you have enough leg room. As your budget is +0.01 every second you could save it, or even just postpone until the tab is focused again.

That may be queuing messages, espacing auto-save, ....
Another suggested measure is to send tasks to do to a worker, which is not throttled.
So it's not as if there is nothing to do.

I'm also expecting that if websockets still avoid throttling for 57, it's not guarantied to not throttle in a post 57 release.

**Comment by sccolbert** [comment]
My hunch is that this throttling won't be around for long. I feel like it's going to break a lot of the internet, such as every music streaming site. Is YouTube going to stop playing in the background? Or will Chrome special case it? I'd say the best approach is to wait-and-see. Nevertheless thanks for bring it to attention!

**Comment by Carreau** [comment]
> My hunch is that this throttling won't be around for long. I feel like it's going to break a lot of the internet, such as every music streaming site.

Yeah, that's one of the reason they already backtracked a lot. AFAIU they implemented other heuristics, like disabling throttling when the audio API is in used. Which in part lead website to play silent sounds in loop to avoid throttling on Chrome Beta.

> Nevertheless thanks for bring it to attention!

No problem, I think we can close as there is no a lot actionable, but at least we know that this may happen.

**Comment by jasongrout** [comment]
> No problem, I think we can close as there is no a lot actionable, but at least we know that this may happen.

Thanks again for bringing it to our attention!
